### PR TITLE
New RFC: What is vs. is not considered a security bug in Wasmtime?

### PR DESCRIPTION
**Pull request by fitzgen**
> Wasmtime has documented processes for reporting, responding to, patching, and disclosing security vulnerabilities. However, the Wasmtime project does not currently define which kinds of bugs are and are not considered security vulnerabilities. This RFC aims to reach consensus on that definition.

### [**Rendered**](https://github.com/fitzgen/rfcs-2/blob/what-is-considered-a-security-bug/accepted/what-is-considered-a-security-bug.md)

### CODE REVIEW

**Comment by cfallin** [comment]
This looks great; thanks a bunch for writing all this out explicitly!

A few thoughts on other things we could include:
- We might want to mention our "on-by-default" policy that we've informally had so far: a security vulnerability is not a CVE (with all the embargo process etc) if it exists only with settings that are off or different by default.
- It might be helpful to dive a bit more into the miscompilation triage process: if we do adopt the definitions here, then e.g. we'd be very suspicious of miscompiles in scalar integer operators, or control-flow operators, because both are used in the lowering/implementation of sandboxing; but we can give an almost blanket statement about SIMD and FP miscompiles not affecting sandboxing (as long as we don't, say, decide to vectorize our address computations).
- Would it make sense to include examples of past CVEs, and of past bugs that are maybe security-adjacent but are not CVEs (or would not be under this proposed definition), and a few sentences about why?

File: accepted/what-is-considered-a-security-bug.md
Comment (tschneidereit) [inline accepted/what-is-considered-a-security-bug.md:175]:
```suggestion
* Are there other kinds of bugs that we don't consider security vulnerabilities
```

**Comment by alexcrichton** [comment]
Another possible angle to consider here is relation to the tier system we have for Wasmtime. For example a sandbox escape in riscv64 is probably not a CVE at this time, but for x86_64 definitely is. Would aarch64 though given its tier 2 status? (I would say probably yes but perhaps this is a special case). That could provide a framework for saying that the pooling allocator is for example tier 1 (despite not being explicitly listed as so) but the component model is "approaching tier 2" sort of given its prior instability. Mostly just food-for-thought.

One additional possible idea that @tschneidereit brought up is that we could consider compiling Cranelift itself to wasm and use that as a sandbox for compilations. For example we could have a `sandbox-compiler` feature on the Wasmtime crate which gives you APIs such as `Config::cranelift_timeout`. That would easily catch panics as recoverable errors, at the cost of compilation performance. That may help provide a useful valve to differentiate between cases of where the compiler is for example process-sandboxed and those where the process sandbox isn't sufficient.

**Comment by cfallin** [comment]
> One additional possible idea that @tschneidereit brought up is that we could consider compiling Cranelift itself to wasm and use that as a sandbox for compilations. For example we could have a `sandbox-compiler` feature on the Wasmtime crate which gives you APIs such as `Config::cranelift_timeout`. That would easily catch panics as recoverable errors, at the cost of compilation performance. That may help provide a useful valve to differentiate between cases of where the compiler is for example process-sandboxed and those where the process sandbox isn't sufficient.

This is a really really cool idea and a nice way to consolidate all of our trusted sandboxing logic into one place. It'd be quite a bit of build-system hacking (invoke a sub-build, `include_bytes!` the `cranelift.wasm` result) and marshalling (the interface between Wasmtime and Cranelift is fairly large; I guess all of `wasmtime compile` technically would need to be included, then we have a `.wasm` -> `.cwasm` pure function) but seems feasible.

**Comment by fitzgen** [comment]
Regarding the "on-by-default" policy: I feel like this is an informal proxy for whether something is tier 1 or not, and now that we've taken the time to define tiers, I think we should stick with just referencing tiers and put the "on-by-default" stuff to rest.

**Comment by fitzgen** [comment]
> One additional possible idea that @tschneidereit brought up is that we could consider compiling Cranelift itself to wasm and use that as a sandbox for compilations.

I agree: this is a really neat idea!

That said, I fear that it will also muddy the waters a bit as to whether a miscompilation is a security bug, since a miscompilation can now affect the whole compiler, and makes it hard to then reason about the resulting `cranelift.wasm`'s compile results and their correctness.

In this very RFC we say

> * Wasmtime embedders should never blindly trust values from the guest &mdash; no
>   matter how trusted the guest program is, even if it was written by the
>   embedders themselves &mdash; and should always validate these values before
>   performing unsafe operations on behalf of the guest.

and I don't know how we would validate `.cwasm`s produced by a miscompiled `cranelift.wasm` for correctness from this perspective.

Basically, instead of looking at just the transitive closure of (1) which instructions are used during bounds checks / other sandboxing-related operations and also (2) the instructions produced by potential optimizations on those first instructions, etc... to determine whether a miscompile is a potential sandbox escape or security bug, we would have to look at the whole of `cranelift.wasm`. And that is so big I don't think we could actually do it in practice.

**Comment by alexcrichton** [comment]
Hm that's a good point yeah, but thinking about this I feel like that section may be a bit too strongly worded. If we can't trust ourselves compiling Cranelift to wasm then that feels wrong to me. This feels analagous to an embedder running a trusted program over untrusted input - there should still, in my opinion, be a degree of trust between Wasmtime and the embedder that the produced value from that computation is still valid. I think it's reasonable to say that the memory safety of the embedder should not hinge upon the correctness of the computation, but it also doesn't feel like going to the other extreme of saying that the output is a totally untrusted value as well.

I suppose personally the way I think about this is that if a miscompilation is found but is not CVE worth due to not breaking the sandbox it's still up to individual embedders to determine whether that affects them or not. For example if we were to compile Cranelift to wasm we would need to evaluate if, after fixing the issue, the output of Cranelift changed at all. If it didn't then that's not a CVE for us, but if it did then we would need a CVE for ourselves about the result of the miscompile, not necessarily the miscompile itself.

Also to be clear I don't think this is the time and/or place to debate about compiling Cranelift to wasm, but I do think it's useful in this context as a concrete example of something to talk about. So I think I would personally suggest rephrasing that section of the RFC to:

> * Wasmtime embedders should not hinge their memory safety guarantees on outputs of the guest, no matter how trusted the guest program is. Values should always be validated before performing unsafe operations on behalf of the guest. Miscompilations should be monitored in the Wasmtime repository (TODO: how would one do this?) and evaluated if this affects how a guest's output is interpreted.

or something like that.

**Comment by cfallin** [comment]
The Cranelift-in-a-Wasm idea does seem pretty relevant to me at least, in the sense that its hypothetical possibility relieves some pressure that might otherwise be felt from the "compiler runaway is a CVE" viewpoint. One concrete thought on correctness of the Cranelift compilation w.r.t. whomever compiles Cranelift is: the status quo today is that Cranelift is compiled by rustc and LLVM. Do we trust that both are bug-free, and issue CVEs for all correctness bugs otherwise? If not, then this wouldn't be a degradation, really.

**Comment by cfallin** [comment]
And actually, now that I think of that, another interesting thing to add to the RFC might be a "policies of peers" section, for reference. E.g., what do SpiderMonkey, V8, LLVM, gcc, HotSpot JVM, ... consider CVEs or security-critical issues?

**Comment by fitzgen** [comment]
@alexcrichton 

> Wasmtime embedders should not hinge their memory safety guarantees on outputs of the guest, no matter how trusted the guest program is.

But... we do hinge memory safety guarantees on the outputs of Cranelift :-p

---------

@cfallin

> the status quo today is that Cranelift is compiled by rustc and LLVM. Do we trust that both are bug-free, and issue CVEs for all correctness bugs otherwise? If not, then this wouldn't be a degradation, really.

True, but we're specifically trusting LLVM compiling to x64 right now, and `cranelift.wasm` would require trusting LLVM compiling to Wasm. The x64 backend is definitely used more than the Wasm backend, but maybe this is just splitting hairs at this point.

**Comment by alexcrichton** [comment]
Personally I would still say that if we can't trust Cranelift compiled to wasm then something isn't aligned right and we need to think about this some more. LLVM-compiling-to-wasm is a good point that it's not on the "critical path" today for Cranelift itself, but it is for just about everyone using Wasmtime because nearly all guest programs had LLVM involved at some point.

I personally still feel the way to think about a possible compiler bug in Cranelift is highly contextual. I think it's reasonable to blanket say that by default they're not considered CVEs unless it leads to a sandbox escape, but we can't escape that whether or not it's a security issue depends on the context. For example we would need to evaluate how it affected cranelift-compiled-to-wasm. SIMD bugs would for example likely not affect this since we don't use much SIMD in Cranelift, and even a floating-point bug may not be that bad as it might just affect some constant calculations as opposed to address layout. An integer computational issue though might affect compiled output which could be a CVE for us.

I would argue that for us we're not blindly trusting the output of Cranelift. We've got efforts to verify Cranelift and it's output so we're doing everything we can to make it suitable for safety-critical use cases, which I would argue double verifies it compiling itself.

**Comment by fitzgen** [comment]
One idea a couple of us talked about internally was whether we could transparently spawn a child process and lock it down with seccomp filters. Seems like this could potentially be a smaller diff than putting all of compilation into Wasm, would still benefit from things like threading, potentially involve less refactoring, and doesn't involve philosophical conversations about how much we trust Cranelift-in-Wasm.

This seems very doable for the CLI, but not really for the library, since we don't have an executable to exec. Could fork/clone without exec, but then we inheret too much from the parent process.

Anyways, this is all starting to get on a bit of a tangent away from the RFC, so maybe if we want to continue this discussion we should open a dedicated issue about sandboxing compilation.

**Comment by alexcrichton** [comment]
Right yes sorry I tried to clarify above, but I agree that this isn't the place to get sidetracked on the specifics of compiling Cranelift to wasm. That being said I still feel that if we aren't comfortable compiling Cranelift to wasm given the advice here then that feels like misaligned incentives and should reword here.

**Comment by ulan** [comment]
Thank you @fitzgen for making the definitions clear. I am happy to hear that DoS during execution is considered a security bug. It would be great to have stronger guarantees for Winch and also treat DoS during Winch compilation as a security bug.

> the Wasmtime embedders might architect their application to compile Wasm in a separate OS process. 

We were thinking about using `Module::serialize()` in a separate compilation process and then pass the serialized binary to the execution process, which would then use `Module::deserialize()`. Is this what you had in mind here as well? Or is there another API we should use instead?

I wonder if `Module::deserialize()` is guaranteed to be resilient and not have the Cranelift DoS issues?

> Cranelift compilation isn't architected for interruption and adding it after the fact would be quite a large undertaking and would probably introduce new performance overhead.

The point about difficulty of limiting the compilation time makes a lot sense. Have you considered stronger guarantees for crashes and panics?

**Comment by fitzgen** [comment]
@ulan 

> It would be great to have stronger guarantees for Winch and also treat DoS during Winch compilation as a security bug.

That could make sense as an eventual goal for Winch. I'd want to get buy in from @saulecabrera before making those kinds of commitments, as he is the primary maintainer of Winch. It is also an area where, if your organization needs these properties, it should probably commit engineering resources towards implementation and maintenance. All that said, Winch is still incomplete, and isn't tier one yet, and therefore we won't currently be making CVEs and doing the full security process for bugs in it yet.

> We were thinking about using `Module::serialize()` in a separate compilation process and then pass the serialized binary to the execution process, which would then use `Module::deserialize()`. Is this what you had in mind here as well? Or is there another API we should use instead?
> 
> I wonder if `Module::deserialize()` is guaranteed to be resilient and not have the Cranelift DoS issues?

Module serialization and deserialization doesn't invoke the compiler, and therefore doesn't suffer from the potential DoS pitfalls that Cranelift has. But just invoking `module.serialize()` in a separate process is also therefore not that useful. The key thing you would want to isolate in a different process would be `Module::new` and `Engine::precompile_module` calls.

> The point about difficulty of limiting the compilation time makes a lot sense. Have you considered stronger guarantees for crashes and panics?

We haven't spent a lot of time distinguishing between different DoS vectors and whether we want to provide stronger guarantees about one vs another. I personally don't think it makes a ton of sense to treat them differently w.r.t. issuing CVEs and our security process because it doesn't ultimately change the story for embedders, as far as I can tell.

Additionally, for panics at least, this is also an area where Cranelift hasn't traditionally been engineered for this goal from the start. There are places where Cranelift will panic on things like too-big input, as you've discovered previously. It is possible we could make stronger guarantees here eventually, but it would require someone auditing Cranelift for these sorts of panics, fixing anything they found, adding testing/fuzzing infrastructure to check for this stuff in the future, and maintaining these properties as things move forward. Again, if this is something your organization feels strongly about, this is an area it could devote resources towards.

All that said, note that all of crashes, panics, infinite loops, OOMs, and etc... are still definitely considered bugs! We want to fix any of these issues if they pop up. And we can even backport fixes to the two most recent releases, like we do for security bugs, if there is demand. It's just a matter of for which things do we commit to doing our whole security process, emailing the security announcement list, issuing CVEs, and all that.

**Comment by saulecabrera** [comment]
@ulan @fitzgen 

My expectation is that Winch should align with Cranelift in terms of guarantees where applicable. So I support the idea of making this expectation explicit in this RFC as an _eventual_ goal for Winch; as Nick mentions above, this expectation is bound to Winch reaching Tier 1, according to [Wasmtime's Tiers of Support.](https://docs.wasmtime.dev/stability-tiers.html#tier-1---production-ready)

**Comment by ulan** [comment]
Thank you @fitzgen and @saulecabrera!

> Module serialization and deserialization doesn't invoke the compiler

This is great! I guess it also means that if we move compilation to a separate process and rely on serialize/deserialize, then that will work with Winch as well.

> So I support the idea of making this expectation explicit in this RFC as an eventual goal for Winch

Sound good, thanks! There might be some design choices that might make achieving this goal either more expensive or cheaper in the future. For example, one issue with Cranelift currently is that some functions panic instead of returning `Result`. Fixing that might require larger refactorings. It would be great if Winch would lean towards not panicking early on. If there is some specific work we could contribute to move in that direction, I'd be happy to discuss it.

> And we can even backport fixes to the two most recent releases, like we do for security bugs, if there is demand.

Yes, this would be super helpful for us until Winch is ready. What would also help us is if you would give us heads up before publishing a fix for a compiler crash/panic/DoS. It is probably not possible in all case, but even a best-effort process would be great. Would you be open to that?

**Comment by fitzgen** [comment]
I forgot that this never merged!

In the time since discussion died down, Wasmtime has gained the ability to build itself just the compiler and without the runtime, and this build itself can be compiled to Wasm, and run in a build of Wasmtime that has the runtime (and may or may not have the compiler). This technique can be used to resolve denial-of-service bugs during compilation, for those who need that. Thanks @adambratschikaye for enabling runtime-free builds!

That seemed like it was the only big open item. Given that, I'm going move that we merge this RFC!

--------

# Motion to Finalize

**Disposition: Merge**

As always, details on the RFC process can be found here: https://github.com/bytecodealliance/rfcs/blob/main/accepted/rfc-process.md#making-a-decision-merge-or-close

**Review by alexcrichton — APPROVED** [review]

**Review by pchickey — APPROVED** [review]

**Review by cfallin — APPROVED** [review]

**Comment by fitzgen** [comment]
As there have been signoffs from at least two people representing different stakeholder organizations, this RFC is entering its 10 day

# Final Comment Period

and the last day to raise objections before this can merge is 2024-02-16.

**Comment by fitzgen** [comment]
Seeing as no objections have been raised during the final comment period, I'm going to go ahead and merge this. Thanks everyone!

**Pull request merged**
